presto-spark-base/src/main/java/com/facebook/presto/spark/PrestoSparkQueryExecutionFactory.java

```diff
@@ -270,11 +258,20 @@ private PrestoSparkQueryExecution(
             queryCompletedEvent(Optional.empty());
 
             ConnectorSession connectorSession = session.toConnectorSession();
-            return resultRdd.stream()
-                    .map(Tuple2::_2)
-                    .map(this::deserializePage)
-                    .flatMap(page -> getPageValues(connectorSession, page, rootFragment.getTypes()).stream())
-                    .collect(toList());
+            List<Type> types = rootFragment.getTypes();
+            ImmutableList.Builder<List<Object>> result = ImmutableList.builder();
+            for (Tuple2<Integer, PrestoSparkRow> tuple : resultRdd) {
+                PrestoSparkRow row = tuple._2;
+                SliceInput sliceInput = new BasicSliceInput(Slices.wrappedBuffer(row.getBytes(), 0, row.getLength()));
+                ImmutableList.Builder<Object> columns = ImmutableList.builder();
+                for (Type type : types) {
+                    BlockBuilder blockBuilder = type.createBlockBuilder(null, 1);
+                    blockBuilder.deserializePosition(sliceInput);
```

Comment:
Have you considered adding something like an overload to `getObjectValue` that takes in a `sliceInput`, to avoid having to create the block builder per cell?

Reply:
This is only used to produce the final output to the client, which is not expected to be large.